---
title: "03_PCA_DESeq2"
format:
  html: default
editor: visual
---

```{r}
library(tidyverse)
library(DESeq2)
library(fgsea)
library(msigdbr)
```

### Step 1: Creating the DESeqDataSet object (Organizing the data)

The first step is to organize the data (gene count matrix + sample meta data) into a DESeqDataSet. This can be done using the DESeqDataSetFromMatrix() function, which is described in the DESeq2 vignettte in the “Count matrix input” section.

```{r}

# 0. safety: don't overwrite originals
melanoma_meta_orig <- melanoma_metadata
melanoma_counts_orig <- melanoma_gene_count

melanoma_metadata <- as.data.frame(melanoma_metadata)

# 1. Set the sample IDs as rownames of the metadata
rownames(melanoma_metadata) <- melanoma_metadata$sample_id

# 2. Check alignment between metadata and count matrix
all(rownames(melanoma_metadata) %in% colnames(melanoma_gene_count))   # should be TRUE
all(rownames(melanoma_metadata) == colnames(melanoma_gene_count))    # TRUE if already in same order

# 3. If not same order, fix it:
melanoma_gene_count <- melanoma_gene_count[, rownames(melanoma_metadata)]

# 4. Build the DESeq2 dataset
dds_melanoma <- DESeqDataSetFromMatrix(
  countData = melanoma_gene_count,
  colData   = melanoma_metadata,
  design    = ~ Groups
)
dds_melanoma
```

### Step 2: PCA (Principal Component Analysis)

```{r}
vsd_melanoma <- rlog(dds_melanoma)   # variance-stabilizing transform
p <- plotPCA(vsd_melanoma, intgroup = "Groups")
ggsave(filename = file.path(plots_dir, "PCA_melanoma.png"), plot = p, width = 8, height = 6, dpi = 300)
```

### Step 3: Testing for differential expression

```{r}
dds_melanoma <- DESeq(dds_melanoma)
deseq2_res_melanoma <- results(dds_melanoma)
summary(deseq2_res_melanoma)
sum(deseq2_res_melanoma$padj < 0.05, na.rm=TRUE) # number of significant genes
```

### Now repeat for the bladder dataset:

```{r}

bladder_meta_orig <- bladder_metadata
bladder_counts_orig <- bladder_gene_count

bladder_metadata <- as.data.frame(bladder_metadata)

# 1. Set the sample IDs as rownames of the metadata
rownames(bladder_metadata) <- bladder_metadata$sample_id

all(rownames(bladder_metadata) %in% colnames(bladder_gene_count))   # should be TRUE
all(rownames(bladder_metadata) == colnames(bladder_gene_count))    # TRUE if already in same order

bladder_gene_count <- bladder_gene_count[, rownames(bladder_metadata)]

dds_bladder <- DESeqDataSetFromMatrix(
  countData = bladder_gene_count,
  colData   = bladder_metadata,
  design    = ~ Groups
)
dds_bladder

vsd_bladder <- rlog(dds_bladder)   # variance-stabilizing transform
p <- plotPCA(vsd_bladder, intgroup = "Groups")

dds_bladder <- DESeq(dds_bladder)
deseq2_res_bladder <- results(dds_bladder)
summary(deseq2_res_bladder)
sum(deseq2_res_bladder$padj < 0.05, na.rm=TRUE) # number of significant genes

ggsave(filename = file.path(plots_dir, "PCA_bladder.png"), plot = p, width = 8, height = 6, dpi = 300)
```

### **Gene Set Enrichment Analysis (Specifically FCS)**

High PC1 variance for both datasets and straightforward study design suggest no obvious confounders, so we did not correct for confounders.

First, let's tackle the melanoma dataset:

```{r}
# 1. Gene sets (Biological Process, Ensembl IDs)
BP_df <- msigdbr(species = "human", category = "C5", subcategory = "BP")
BP_list <- split(BP_df$ensembl_gene, BP_df$gs_name)

# 2. Make ranked stats vector (use Ensembl IDs as names)
myStats_melanoma <- deseq2_res_melanoma[,"stat"]
names(myStats_melanoma) <- rownames(deseq2_res_melanoma)

# 3. Remove version suffix
clean_names_melanoma <- sub("\\.\\d+$", "", names(myStats_melanoma))
names(myStats_melanoma) <- clean_names_melanoma

# 4. Clean up
myStats_melanoma <- myStats_melanoma[!is.na(myStats_melanoma)]
myStats_melanoma <- sort(myStats_melanoma, decreasing = TRUE)

# 5. Run fgsea
fgseaRes_melanoma <- fgseaMultilevel(pathways = BP_list, stats = myStats_melanoma,
                            minSize = 15, maxSize = 500, scoreType = "std")
# 6. Results
sum(fgseaRes_melanoma$padj < 0.05)
head(fgseaRes_melanoma[order(fgseaRes_melanoma$padj), ], 10)
```

Now the bladder:

```{r}
# 2. Make ranked stats vector (use Ensembl IDs as names)
myStats_bladder <- deseq2_res_bladder[,"stat"]
names(myStats_bladder) <- rownames(deseq2_res_bladder)

# 3. Remove version suffix
clean_names_bladder <- sub("\\.\\d+$", "", names(myStats_bladder))
names(myStats_bladder) <- clean_names_bladder

# 4. Clean up
myStats_bladder <- myStats_bladder[!is.na(myStats_bladder)]
myStats_bladder <- sort(myStats_bladder, decreasing = TRUE)

# 5. Run fgsea
fgseaRes_bladder <- fgseaMultilevel(pathways = BP_list, stats = myStats_bladder,
                            minSize = 15, maxSize = 500, scoreType = "std")
# 6. Results
sum(fgseaRes_bladder$padj < 0.05)
head(fgseaRes_bladder[order(fgseaRes_bladder$padj), ], 10)
```

Now let’s sort by significance, pick the top pathway and plot enrichment.

Melanoma

```{r}
# sort significant results
fg_sig_melanoma <- fgseaRes_melanoma %>%
  arrange(padj) %>%
  filter(padj < 0.05)

# pick the top pathway
chosen_pathway_melanoma <- fg_sig_melanoma$pathway[1]

# plot enrichment
p <- plotEnrichment(BP_list[[chosen_pathway_melanoma]], myStats_melanoma) +
  ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_melanoma))

# extract first 3 leading-edge genes
leading_melanoma <- fg_sig_melanoma$leadingEdge[[1]]
first_genes_melanoma <- head(leading_melanoma, 3)
first_genes_melanoma

ggsave(filename = file.path(plots_dir, "DESeq2_enrichment_melanoma.png"), plot = p, width = 8, height = 6, dpi = 300)
```

Bladder

```{r}
# sort significant results
fg_sig_bladder <- fgseaRes_bladder %>%
  arrange(padj) %>%
  filter(padj < 0.05)

# pick the top pathway
chosen_pathway_bladder <- fg_sig_bladder$pathway[1]

# plot enrichment
p <- plotEnrichment(BP_list[[chosen_pathway_bladder]], myStats_bladder) +
  ggplot2::labs(title = paste0("Enrichment: ", chosen_pathway_bladder))

# extract first 3 leading-edge genes
leading <- fg_sig_bladder$leadingEdge[[1]]
first_genes_bladder <- head(leading, 3)
first_genes_bladder

ggsave(filename = file.path(plots_dir, "DESeq2_enrichment_bladder.png"), plot = p, width = 8, height = 6, dpi = 300)
```

**Non-directional on melanoma:**

```{r}
stats_abs_melanoma <- abs(myStats_melanoma)

fg_nondir_melanoma <- fgseaMultilevel(
  pathways = BP_list,
  stats = stats_abs_melanoma,
  minSize = 15,
  maxSize = 500,
  scoreType = "pos"
)

fg_nondir_melanoma <- fg_nondir_melanoma[order(fg_nondir_melanoma$padj), ]
head(fg_nondir_melanoma, 10)
```

```{r}
fg_volcano_melanoma <- fgseaRes_melanoma %>%
  mutate(sig = padj < 0.05,
         mlog10padj = -log10(padj))

to_label <- fg_volcano_melanoma %>% arrange(padj) %>% slice_head(n = 10)

library(ggrepel)
p <- ggplot(fg_volcano_melanoma, aes(x = NES, y = mlog10padj, color = sig)) +
  geom_point(alpha = 0.8) +
  geom_hline(yintercept = -log10(0.05), linetype = 2) +
  geom_text_repel(data = to_label,
                  aes(label = stringr::str_trunc(pathway, 50)),
                  max.overlaps = 30, size = 3) +
  scale_color_manual(values = c("FALSE" = "grey80", "TRUE" = "#1f77b4"),
                     labels = c("n.s.", "< 0.05")) +
  labs(x = "NES",
       y = expression(-log[10](padj)),
       color = "Significant",
       title = "Volcano plot of gene-set enrichment for the melanoma dataset") +
  theme_minimal(base_size = 11)
p
ggsave(filename = file.path(plots_dir, "DESeq2 Volcano plot (Melanoma).png"), plot = p, width = 8, height = 6, dpi = 300)
```

**Non-directional on bladder:**

```{r}
stats_abs_bladder <- abs(myStats_bladder)

fg_nondir_bladder <- fgseaMultilevel(
  pathways = BP_list,
  stats = stats_abs_bladder,
  minSize = 15,
  maxSize = 500,
  scoreType = "pos"
)

fg_nondir_bladder <- fg_nondir_bladder[order(fg_nondir_bladder$padj), ]
head(fg_nondir_bladder, 10)
```

```{r}
fg_volcano_bladder <- fgseaRes_bladder %>%
  mutate(sig = padj < 0.05,
         mlog10padj = -log10(padj))

to_label <- fg_volcano_bladder %>% arrange(padj) %>% slice_head(n = 10)

p <- ggplot(fg_volcano_bladder, aes(x = NES, y = mlog10padj, color = sig)) +
  geom_point(alpha = 0.8) +
  geom_hline(yintercept = -log10(0.05), linetype = 2) +
  geom_text_repel(data = to_label,
                  aes(label = stringr::str_trunc(pathway, 50)),
                  max.overlaps = 30, size = 3) +
  scale_color_manual(values = c("FALSE" = "grey80", "TRUE" = "#1f77b4"),
                     labels = c("n.s.", "< 0.05")) +
  labs(x = "NES",
       y = expression(-log[10](padj)),
       color = "Significant",
       title = "Volcano plot of gene-set enrichment for the bladder dataset") +
  theme_minimal(base_size = 11)
p
ggsave(filename = file.path(plots_dir, "DESeq2 Volcano plot (Bladder).png"), plot = p, width = 8, height = 6, dpi = 300)
```

## Cleaning environment

```{r}
rm(
  melanoma_meta_orig, melanoma_counts_orig,
  bladder_meta_orig, bladder_counts_orig,
  BP_df,
  myStats_melanoma, myStats_bladder,
  clean_names_melanoma, clean_names_bladder,
  fg_sig_melanoma, fg_sig_bladder,
  stats_abs_melanoma, stats_abs_bladder,
  fg_nondir_melanoma, fg_nondir_bladder,
  fg_volcano_melanoma, fg_volcano_bladder,
  to_label,
  chosen_pathway_melanoma, chosen_pathway_bladder,
  leading_melanoma, leading,
  first_genes_melanoma, first_genes_bladder,
  p
)
gc()

```
