---
title: "04_DEXSeq"
format: html
---

# DEXSeq

## Load libraries

```{r}
library(DEXSeq)
library(tidyverse)
library(dplyr)
library(tibble)
library(ggplot2)
```

## Metadata preparation

```{r}
# MELANOMA METADATA
melanoma_metadata <- melanoma_metadata %>%
  dplyr::select(sample = sample_id,
                condition = Groups) %>%
  dplyr::mutate(
    condition = factor(condition,
                       levels = c("untreated", "treated"))
  ) %>%
  as.data.frame()

rownames(melanoma_metadata) <- melanoma_metadata$sample
melanoma_metadata$sample <- NULL

stopifnot( all(rownames(melanoma_metadata) == colnames(melanoma_iso_count)) )
```

```{r}
# BLADDER METADATA
bladder_metadata <- bladder_metadata %>% 
  dplyr::select(sample = sample_id,
                condition = Groups) %>% 
  dplyr::mutate(
    condition = factor(condition,
                       levels = c("untreated", "treated"))
  ) %>% 
  as.data.frame()

rownames(bladder_metadata) <- bladder_metadata$sample
bladder_metadata$sample <- NULL

stopifnot( all(rownames(bladder_metadata) == colnames(bladder_iso_count)) )
```

## Transcript–Gene Mapping and Annotation

The code retrieves and aligns **transcript–gene mapping** information, ensuring that the filtered count matrices for **melanoma** and **bladder** samples match the correct **transcript IDs**.
It then creates **annotated count tables** by joining the mapping data, adding the corresponding **gene IDs** and **gene names** to each **transcript**.

```{r}
#GETTING ISO INFO
iso_gene_map <- isoInfo %>%
  dplyr::select(
    transcript_id,
    gene_id,
    gene_name
  )
head(iso_gene_map)

# mappa trascritto-gene per MELANOMA (matrice filtrata)
mel_txInfo <- iso_gene_map[match(rownames(melanoma_iso_count_filt),
                                 iso_gene_map$transcript_id), ]
stopifnot(all(mel_txInfo$transcript_id == rownames(melanoma_iso_count_filt)))

# mappa trascritto-gene per BLADDER (matrice filtrata)
blad_txInfo <- iso_gene_map[match(rownames(bladder_iso_count_filt),
                                  iso_gene_map$transcript_id), ]
stopifnot(all(blad_txInfo$transcript_id == rownames(bladder_iso_count_filt)))

#CREATE MELANOMA ASSOCIATIONS
mel_counts_annot <- melanoma_iso_count_filt %>%
  as.data.frame() %>%
  tibble::rownames_to_column("transcript_id") %>%
  dplyr::left_join(iso_gene_map, by = "transcript_id") %>%
  dplyr::relocate(gene_id, gene_name, .before = transcript_id)

head(mel_counts_annot)


#CREATE BLADDER ASSOCIATIONS
blad_counts_annot <- bladder_iso_count_filt %>%
  as.data.frame() %>%
  tibble::rownames_to_column("transcript_id") %>%
  dplyr::left_join(iso_gene_map, by = "transcript_id") %>%
  dplyr::relocate(gene_id, gene_name, .before = transcript_id)

head(blad_counts_annot)
```

## DEXSeq analysis

**Dataset Construction and Differential Exon Usage Testing**: This chunk builds the DEXSeq datasets for melanoma and bladder samples, performs normalization and dispersion estimation, and computes differential exon usage results for each condition.

```{r}
dxd_mel <- DEXSeqDataSetFromMatrix(
  countData = melanoma_iso_count_filt,
  colData    = melanoma_metadata,
  design     = ~ sample + exon + condition:exon,
  featureID  = mel_txInfo$transcript_id,
  groupID    = mel_txInfo$gene_id
)

dxd_blad <- DEXSeqDataSetFromMatrix(
  countData  = bladder_iso_count_filt,
  colData    = bladder_metadata,
  design     = ~ sample + exon + condition:exon,
  featureID  = blad_txInfo$transcript_id,
  groupID    = blad_txInfo$gene_id
)

dxd_mel  <- estimateSizeFactors(dxd_mel)
dxd_mel  <- estimateDispersions(dxd_mel)
dxd_mel  <- testForDEU(dxd_mel)
dxd_mel  <- estimateExonFoldChanges(dxd_mel, fitExpToVar = "condition")
dexseq_res_melanoma  <- DEXSeqResults(dxd_mel)
  
dxd_blad <- estimateSizeFactors(dxd_blad)
dxd_blad <- estimateDispersions(dxd_blad)
dxd_blad <- testForDEU(dxd_blad)
dxd_blad <- estimateExonFoldChanges(dxd_blad, fitExpToVar = "condition")
dexseq_res_bladder <- DEXSeqResults(dxd_blad)

```

**Convert DEU Results to Data Frames:**
This chunk converts the DEXSeq results objects for melanoma and bladder into data frames, preparing them for downstream filtering and significance analysis.

```{r}
## MELANOMA ---- CREATE DATAFRAME AND SORT IT BY ADJUSTED P VALUE
dex_mel_df <- as.data.frame(dexseq_res_melanoma)
## BLADDER ---- CREATE DATAFRAME AND SORT IT BY ADJUSTED P VALUE
dex_blad_df <- as.data.frame(dexseq_res_bladder)
```

**Identify Significant Isoforms and Genes:** This chunk filters DEU results by adjusted p-value to quantify the number of significantly affected isoforms and the corresponding unique genes.

```{r}
alpha <- 0.05

## MELANOMA ----
# Significant isoforms (rows)
sig_iso_mel <- dex_mel_df[!is.na(dex_mel_df$padj) & dex_mel_df$padj < alpha, ]
n_sig_isoforms_mel <- nrow(sig_iso_mel)
n_sig_isoforms_mel

# Significant genes (unique groupID)
n_sig_genes_mel <- length(unique(sig_iso_mel$groupID))
n_sig_genes_mel
```

```{r}
alpha <- 0.05

## BLADDER ----
# Significant isoforms (rows)
sig_iso_blad <- dex_blad_df[!is.na(dex_blad_df$padj) & dex_blad_df$padj < alpha, ]
n_sig_isoforms_blad <- nrow(sig_iso_blad)
n_sig_isoforms_blad

# Significant genes (unique groupID)
n_sig_genes_blad <- length(unique(sig_iso_blad$groupID))
n_sig_genes_blad
```

# Plots

## Volcano plot

```{r}
### MELANOMA VOLCANO ----
dex_mel_df$mlog10padj <- -log10(dex_mel_df$padj)
lfc_mel <- dex_mel_df %>% dplyr::select(starts_with("log2fold_")) %>% pull()
p <- ggplot(dex_mel_df, aes(x = lfc_mel, y = mlog10padj)) +
  geom_point(aes(color = padj < 0.05), alpha = 0.7) +
  scale_color_manual(values = c("grey70", "red")) +
  geom_hline(yintercept = -log10(0.05), linetype = 2) +
  labs(
    title = "DEXSeq Volcano Plot – Melanoma",
    x = "log2 Fold Change (Exon usage)",
    y = "-log10(padj)",
    color = "Significant\n(padj < 0.05)"
  ) +
  theme_minimal()

ggsave(filename = file.path(plots_dir, "DEXSeq Volcano Plot (Melanoma).png"), plot = p, width = 8, height = 6, dpi = 300)

### BLADDER VOLCANO ----
dex_blad_df$mlog10padj <- -log10(dex_blad_df$padj)
lfc_blad <- dex_blad_df %>% dplyr::select(starts_with("log2fold_")) %>% pull()
p <- ggplot(dex_blad_df, aes(x = lfc_blad, y = mlog10padj)) +
  geom_point(aes(color = padj < 0.05), alpha = 0.7) +
  scale_color_manual(values = c("grey70", "red")) +
  geom_hline(yintercept = -log10(0.05), linetype = 2) +
  labs(
    title = "DEXSeq Volcano Plot – Bladder",
    x = "log2 Fold Change (Exon usage)",
    y = "-log10(padj)",
    color = "Significant\n(padj < 0.05)"
  ) +
  theme_minimal()

ggsave(filename = file.path(plots_dir, "DEXSeq Volcano Plot (Bladder).png"), plot = p, width = 8, height = 6, dpi = 300)
```

## Ma plot

```{r}
### MELANOMA MA-PLOT ----
p <- plotMA(
  dexseq_res_melanoma,
  main = "DEXSeq MA Plot – Melanoma",
  cex = 0.5
)

ggsave(filename = file.path(plots_dir, "DEXSeq MA Plot (Melanoma).png"), plot = p, width = 8, height = 6, dpi = 300)

### BLADDER MA-PLOT ----
p <- plotMA(
  dexseq_res_bladder,
  main = "DEXSeq MA Plot – Bladder",
  cex = 0.5
)

ggsave(filename = file.path(plots_dir, "DEXSeq MA Plot (Bladder).png"), plot = p, width = 8, height = 6, dpi = 300)
```

## Heatmap

## 
